crates/biome_ungrammar/src/parser.rs

```diff
@@ -108,37 +110,63 @@ fn node(p: &mut Parser) -> Result<()> {
     Ok(())
 }
 
+/// Parse any Rule, the right-hand side of a production. This handles
+/// all of the combinators other than juxtaposition:
+/// 'auto' | Expr | Value
+/// length || color || direction
+/// veritcal && horizontal
 fn rule(p: &mut Parser) -> Result<Rule> {
     if let Some(lexer::Token {
-        kind: TokenKind::Pipe,
+        kind: TokenKind::Pipe | TokenKind::DoubleAmpersand | TokenKind::DoublePipe,
         loc,
     }) = p.peek()
     {
         bail!(
             *loc,
             "The first element in a sequence of productions or alternatives \
-            must not have a leading pipe (`|`)"
+            must not be a combinator (`|`, `||`, or `&&`)"
         );
     }
 
     let lhs = seq_rule(p)?;
-    let mut alt = vec![lhs];
+    let mut rules = vec![lhs];
+    let mut combinator_kind: Option<CombinatorKind> = None;
     while let Some(token) = p.peek() {
-        if token.kind != TokenKind::Pipe {
+        let token_combinator = CombinatorKind::new(&token.kind);
+
+        if matches!(token_combinator, CombinatorKind::NonCombinator) {
             break;
         }
+
+        match combinator_kind {
+            Some(kind) if kind != token_combinator => {
+                bail!(token.loc, "Cannot mix combinators at the same level in a Rule. Use parentheses to specify precedence");
+            }
+            None => combinator_kind = Some(token_combinator),
+            _ => (),
+        }
+
         p.bump()?;
         let rule = seq_rule(p)?;
-        alt.push(rule)
+        rules.push(rule)
     }
-    let res = if alt.len() == 1 {
-        alt.pop().unwrap()
+    let res = if rules.len() == 1 {
+        rules.pop().unwrap()
     } else {
-        Rule::Alt(alt)
+        match combinator_kind {
+            Some(CombinatorKind::DoubleAmpersand) => Rule::UnorderedAll(rules),
+            Some(CombinatorKind::DoublePipe) => Rule::UnorderedSome(rules),
+            Some(CombinatorKind::Pipe) => Rule::Alt(rules),
+            None | Some(CombinatorKind::NonCombinator) => {
+                unreachable!("Matched more than one rule but didn't determine a combinator")
```

Comment:
What should we do in case we mistakenly reach this code path?

Reply:
This should be genuinely unreachable, since we break out of the loop before parsing a second rule if we encounter a non-combinator, and we also bail the parse entirely if we encounter mixed combinators, so I think this is safe as-is. Reaching this state would mean there's both a bug in the parsing logic above it _and_ an unexpected input.